src/bootstrap/suggest.rs

```diff
@@ -0,0 +1,157 @@
+use crate::builder::{Builder, Kind};
+use glob::Pattern;
+use std::{
+    fmt::{self, Display},
+    path::PathBuf,
+    str::FromStr,
+};
+
+/// Suggests a list of possible `x.py` commands to run based on modified files in branch.
+pub fn suggest(build: &Builder<'_>, run: bool) {
+    // Find the commit the current branch split from master.
+    let merge_base = build
+        .config
+        .git()
+        .arg("merge-base")
+        .arg("HEAD")
+        .arg("origin/master")
+        .output()
+        .expect("failed to execute `git`");
+
+    // `git` shouldn't ever give us non-utf8 output.
+    let merge_base = String::from_utf8(merge_base.stdout).expect("`git` gave non-utf8 commit?!");
+
+    // Note that `merge_base` ends in a newline; it must be trimmed.
+    let modified_files = build
+        .config
+        .git()
+        .arg("diff")
+        .arg("--name-only")
+        .arg(merge_base.trim())
+        .output()
+        .expect("failed to execute `git`");
+
+    let modified_files =
+        String::from_utf8(modified_files.stdout).expect("`git` gave non-utf8 commit?!");
```

Comment:
You can get this from [src/bootstrap/format.rs:81-101](https://github.com/rust-lang/rust/blob/master/src/bootstrap/format.rs#L81-L101), and it is a `git` invocation less. Maybe again wait for #105058 to be merged and move it to `src/tools/build-helper/src/git.rs`

Reply:
You should also wait for 106xxx

Reply:
Everything seems quite up in the air. #105058 is getting reverted. I think I'll just wait for this to blow over a bit.